cpp/modmesh/inout/plot3d.cpp

```diff
@@ -0,0 +1,148 @@
+    size_t x_base_idx = 0, y_base_idx = 0, z_base_idx = 0;
+    uint_type idx = 0;
+    uint_type nblocks = 0;
+    uint_type base_idx = 0;
+    // hexahedron has 8 nodes + 1 for store nodes number
+    small_vector<uint_type> nds_temp(9, 0U);
+    // These vector of x, y, z coordinate shift is used to build the hexahedron element
+    // the nodes order of the hexahedron element is clock of the xy face
+    // Referece: https://github.com/nasa/Plot3D_utilities/blob/main/colab/Plot3D_SplitBlocksExample.ipynb
+    small_vector<small_vector<int_type>> shift = {
+        {-1, -1, -1},
+        {-1, 0, -1},
+        {-1, 0, 0},
+        {-1, -1, 0},
+        {0, -1, -1},
+        {0, 0, -1},
+        {0, 0, 0},
+        {0, -1, 0}};
+
+    stream << data;
+    // getting the first line for total blocks number of the p3d mesh file
+    std::getline(stream, line);
+    nblocks = stoul(line);
+
+    m_x_shape.remake(small_vector<size_t>{nblocks}, 0);
+    m_y_shape.remake(small_vector<size_t>{nblocks}, 0);
+    m_z_shape.remake(small_vector<size_t>{nblocks}, 0);
+    m_blk_sizes.remake(small_vector<size_t>{nblocks}, 0);
+
+    // parsing xyz dimension of each block
+    for (auto i = 0; i < nblocks; ++i)
+    {
+        std::getline(stream, line);
+        auto tokens = tokenize(line, "\\d+");
+        m_x_shape(i) = std::stoul(tokens[0]);
+        m_y_shape(i) = std::stoul(tokens[1]);
+        m_z_shape(i) = std::stoul(tokens[2]);
+        m_blk_sizes(i) = m_x_shape(i) * m_y_shape(i) * m_z_shape(i);
+        total_blk_size += m_blk_sizes(i);
+    }
+
+    m_nds.remake(small_vector<size_t>{total_blk_size, 3}, 0);
+
+    // processing the first block first.
+    for (auto k = 0; k < 3; ++k)
+    {
+        for (auto j = 0; j < m_blk_sizes(0); ++j)
+        {
+            if (parsing_q.empty())
+            {
+                std::getline(stream, line);
+                // using regex to parsing the string and split them into float number
+                // also need to consider the scientific notation.
+                auto tokens = tokenize(line, "(-?\\d+(\\.\\d+)?([eE][+-]?\\d+)?|\\b\\d+\\b)");
+                for (auto token : tokens) parsing_q.push(std::stod(token));
+            }
+            m_nds(j, k) = parsing_q.front();
+            parsing_q.pop();
+        }
+    }
+
+    for (auto blk = 1; blk < nblocks; ++blk)
+    {
+        base_idx += m_blk_sizes(blk - 1);
+
+        for (auto k = 0; k < 3; ++k)
+        {
+            for (auto j = 0; j < m_blk_sizes(blk); ++j)
+            {
+                if (parsing_q.empty())
+                {
+                    std::getline(stream, line);
+                    auto tokens = tokenize(line, "(-?\\d+(\\.\\d+)?([eE][+-]?\\d+)?|\\b\\d+\\b)");
+                    for (auto token : tokens) parsing_q.push(std::stod(token));
+                }
+                m_nds(j + base_idx, k) = parsing_q.front();
+                parsing_q.pop();
+            }
+        }
+    }
```

Comment:
I think there may be opportunities to optimize this code because there is some redundant code here.

Reply:
Please leave a code comment to remind the future maintainer.